Upgrade  arrowutils.SortRecord

gernest opened the pull request:
Followup to #628

This commit enhances `arrowutils.SortRecord` to support

- Multi column sort
- Column direction (ascending, descending)
- NullFirst (for both ascending and descending)

Also optimizes `arrowutils.ReorderRecord` to support dictionary columns.

asubiotto commented on pqarrow/arrowutils/sort_test.go:
```diff
@@ -11,73 +11,356 @@ import (
 )
 
 func TestSortRecord(t *testing.T) {
```
I think it would also be great if we could add some benchmarks for these.

gernest replied:
I'm pretty new to `frostdb`, this probably my second or third week on the code base. While I'm comfortatable writing tests, I think I'm not qualified to write benchmarks yet.

I hope someone more knowledgable and better qualified should help with benchmarks. What I'm saying benchmarking is beyond the scope of this PR.

asubiotto replied:
How about now? 😂 I feel you are more than qualified given your recent contributions, but up to you.

gernest replied:
haha! lol! I think I can do it, will make sure we have a benchmark in place.

metalmatze commented on pqarrow/arrowutils/sort.go:
```diff
@@ -2,93 +2,303 @@ package arrowutils
+	mem memory.Allocator,
+	r arrow.Record,
+	columns []int,
+	direction []int,
```
Passing these as structs would be nice!
I was already playing with the idea during the previous PR but it seemed overkill for what it was.

metalmatze commented on pqarrow/arrowutils/sort.go:
```diff
@@ -2,93 +2,303 @@ package arrowutils
+//
+// Comparison is made sequentially by each column. When rows are equal in the
+// first column we compare the rows om the second column and so on and so forth
+// until rows that are not equal are found.
```
Some explanation would be good, yep! 
To be totally honest, to this day I don't get why index columns in arrow aren't unsigned by default too.

gernest replied:
done ff6d120

metalmatze commented on pqarrow/arrowutils/sort.go:
```diff
@@ -2,93 +2,303 @@ package arrowutils
+	if !hasDictionary {
+		res, err := compute.Take(
+			ctx,
+			compute.TakeOptions{BoundsCheck: true},
```
Isn't that the default? 
https://github.com/apache/arrow/blob/go/v14.0.0/go/arrow/compute/expression.go#L448

gernest replied:
Yes, the arrow api is pretty much broken on this regard `compute.DefaultTakeOptions` is a pointer while `compute.Take` accepts a non pointer. Doing pointer derefence for arguments in this case is just not okay.

metalmatze reviewed (commented):
Amazing work! 💯 
Really liking the overall direction you're taking this in!

gernest commented on pqarrow/arrowutils/sort.go:
```diff
@@ -2,93 +2,303 @@ package arrowutils
+		return nil, fmt.Errorf("pqarrow/arrowutils: expected %d nullsFirst got %d", len(columns), len(nullsFirst))
+	}
+
+	build := array.NewInt32Builder(mem)
```
done 2dcf9be

gernest commented on pqarrow/arrowutils/sort.go:
```diff
@@ -2,93 +2,303 @@ package arrowutils
-	// populate indices
-	for i := range indices {
-		indices[i] = int64(i)
+	ms := &multiColSorter{
```
done ca45c04

gernest commented on pqarrow/arrowutils/sort.go:
```diff
@@ -2,93 +2,303 @@ package arrowutils
+	// dictionary column.
+	var hasDictionary bool
+	for i := 0; i < int(r.NumCols()); i++ {
+		_, hasDictionary = r.Column(i).(*array.Dictionary)
```
done 87436e1

gernest commented:
@metalmatze @asubiotto thanks for the feedback, PTAL. I addressed all the comments.

Ignore `fix lint` commits, the linter action was generating partial recommendations, I had to fight for my life to make it happy.

asubiotto commented on pqarrow/arrowutils/sort.go:
```diff
@@ -2,93 +2,306 @@ package arrowutils
+	case Descending:
+		return 1
+	default:
+		return 0
```
I would panic in this default case since it doesn't make sense

gernest replied:
done c00f9d4

asubiotto commented on pqarrow/arrowutils/sort.go:
```diff
@@ -2,93 +2,306 @@ package arrowutils
+// first column we compare the rows om the second column and so on and so forth
+// until rows that are not equal are found.
+func SortRecord(
+	indicesBuilder *builder.OptInt32Builder,
```
I don't like the API of having to instantiate a builder and then passing it in to `SortRecord`. Any reason you did it this way? Can we remove and instantiate within the `SortRecord` function?

gernest replied:
done 18f38f1

asubiotto commented on pqarrow/arrowutils/sort.go:
```diff
@@ -2,93 +2,306 @@ package arrowutils
+}
+
+// SortRecord sorts given arrow.Record by columns.
+// Returns  *array.Int32 of  indices to sorted rows or record r.
```
nit while you're here: extra spaces in comment

gernest replied:
done c087044

asubiotto commented on pqarrow/builder/optbuilders.go:
```diff
@@ -637,6 +638,22 @@ func (b *OptInt32Builder) Set(i int, v int32) {
+
+// UnsafeValue returns value at index i without bounds check.
+func (b *OptInt32Builder) UnsafeValue(i int) int32 {
+	return b.data[i]
```
I don't think we need a bounds check in `Value` since this is performed implicitly by the runtime (just that the panic would be different. We can probably remove this `UnsafeValue` and remove the `if` from `Value`. What do you think? cc @metalmatze since I think you introduced this code.

metalmatze replied:
Not exactly sure anymore. Let's see how it goes now that it's merged. Happy to revise in the future.

asubiotto replied:
The problem is that we now have duplicate methods that do the same thing (and one version seems to have a bug with setting nulls IIRC from another comment on this PR). We should probably fix this sooner rather than later.

gernest replied:
@metalmatze @asubiotto I created a PR for this #680 we can move this discussion there.

asubiotto commented on pqarrow/builder/optbuilders.go:
```diff
@@ -637,6 +638,22 @@ func (b *OptInt32Builder) Set(i int, v int32) {
+// UnsafeSet sets v at index i without bounds check.
+func (b *OptInt32Builder) UnsafeSet(i int, v int32) {
+	b.data[i] = v
+	bitutil.SetBit(b.validityBitmap, i)
```
Interesting, is the fact that we don't set the null bitmap in `Set` a bug? Same idea here as below in `UnsafeValue`: I don't think the application-level bounds check is necessary so we could probably merge

gernest commented:
@asubiotto I addressed your comments PTAL.

**NOTE** . I have left `*OptInt32Builder` api comments because changing api or api expectation is sensitive and they will need more context and input from @metalmatze . I can handle whatever feedback in followup PR.  I have also decided to leave benchmark out.

asubiotto commented on pqarrow/arrowutils/sort.go:
```diff
@@ -2,93 +2,306 @@ package arrowutils
+}
 
-	return indicesBuilder.NewInt64Array(), nil
+func takeDictColumn(ctx context.Context, a *array.Dictionary, idx int, arr []arrow.Array, indices *array.Int32) func() error {
```
nit: how about simply returning an error instead of a callback in all these `take` functions?? You'd have to call this within a `func() error { }` block in the `Go` above, but I think it's nicer given that it gives the option to directly execute this function

gernest replied:
done 7be2e2b

asubiotto commented on pqarrow/builder/optbuilders.go:
```diff
@@ -688,3 +705,9 @@ func (b *OptInt32Builder) ResetToLength(n int) {
 	b.validityBitmap = resizeBitmap(b.validityBitmap, n)
 }
+
+func (b *OptInt32Builder) ReserveToLength(n int) {
```
can we just call this `Reserve` to be more in line with the arrow API?

gernest replied:
done c236674

asubiotto reviewed (commented):
Just a couple of nits, but this is looking good!

gernest commented:
@asubiotto  thanks for the review. I addressed all comments PTAL.

Hopeful we put this into a practical use. Areas I think it will have impact

- logictest : we can now build records directly and skip parquet rows construction

- in parca

https://github.com/parca-dev/parca/blob/4f536249850d27a79b365d3d796ca8ff692896d5/pkg/parcacol/ingest.go#L63-L107
Now we have a generic builder that constructs arrow records directly, sorting was the only missing piece.

asubiotto reviewed (approved):
Absolutely, this is hugely impactful work, thanks for your contribution!

metalmatze commented:
Fantastic work! :100:
